File: python/chatbot_example.py
import os
from openai import OpenAI

# Make sure to set your environment variables accordingly
ENDPOINT = os.getenv("AI_ENDPOINT")
DEPLOYMENT_NAME = os.getenv("DEPLOYMENT_NAME")
API_KEY = os.getenv("AI_API_KEY")

class ChatBot:
    def __init__(self, system=""):
        self.number_of_prompts_sent = 0  # Initialize prompt counter
        self.system = system
        self.messages = []
        if self.system:
            self.messages.append({"role": "system", "content": system})

    def __call__(self, message):
        self.messages.append({"role": "user", "content": message})
        result = self.execute()
        self.messages.append({"role": "assistant", "content": result})
        return result

    def execute(self):
        self.number_of_prompts_sent += 1
        print("\n-- Prompt", self.number_of_prompts_sent, "--")
        print("\n=== FULL PROMPT SENT TO MODEL ===")  
        # Show the complete prompt being sent to the model
        for i, msg in enumerate(self.messages, 1):
            # print(f"Message {i} ({msg['role']}):")
            print(f"  {msg['content']}")
            print()

        client = OpenAI(
            base_url=f"{ENDPOINT}",
            api_key=API_KEY
        )

        response = client.chat.completions.create(
            model=DEPLOYMENT_NAME,
            messages=self.messages
        )
        # Uncomment this to print out token usage each time, e.g.
        # CompletionUsage(completion_tokens=27, prompt_tokens=225, total_tokens=252, completion_tokens_details=CompletionTokensDetails(accepted_prediction_tokens=0, audio_tokens=0, reasoning_tokens=0, rejected_prediction_tokens=0), prompt_tokens_details=PromptTokensDetails(audio_tokens=0, cached_tokens=0))
        # print(response.usage)
                
        print("=== END OF PROMPT ===\n")
        return response.choices[0].message.content
        

def show_examples():
    """Run some example queries to demonstrate the system"""
    print("=" * 80)
    print("CHATBOT EXAMPLES - Simple Q&A")
    print("=" * 80)
    
    examples = [
        "What states does Ohio share borders with?",
        "Calculate the square root of 256.",
        "How many r's are in strawbery?",
        "Who was the first president of the United States?"
    ]
    
    for i, example in enumerate(examples, 1):
        print(f"\n--- Example {i} ---")
        print(example)


def query(question, bot=None): 
    result = bot(question)
    return result

def interactive_loop():
    # Main interactive loop for user queries
    print("=" * 80)
    print("INTERACTIVE CHATBOT - Have a conversation!")
    print("Type your questions and press Enter. Use Ctrl+C to exit.")
    print("=" * 80)
    
    bot = ChatBot()  # Create one bot instance to maintain conversation history

    try:
        while True:
            user_query = input("\nYour question: ").strip()
            if user_query:  # Only process non-empty queries
                result = query(user_query, bot)
                print(f"Answer: {result}")
            else:
                print("Please enter a question.")
    except KeyboardInterrupt:
        print("\n\nGoodbye! Thanks for chatting!")
    except EOFError:
        print("\n\nGoodbye! Thanks for chatting!")

if __name__ == '__main__':
    # First show examples
    show_examples()
    
    # Then start interactive mode
    interactive_loop()

File: node/chatbot_example.ts
import OpenAI from 'openai';
import * as readline from 'readline';

// Make sure to set your environment variables accordingly
const ENDPOINT = process.env.AI_ENDPOINT;
const DEPLOYMENT_NAME = process.env.DEPLOYMENT_NAME;
const API_KEY = process.env.AI_API_KEY;

interface Message {
    role: 'system' | 'user' | 'assistant';
    content: string;
}

class ChatBot {
    private numberOfPromptsSent: number = 0;
    private system: string;
    private messages: Message[] = [];
    private client: OpenAI;

    constructor(system: string = "") {
        this.system = system;
        this.messages = [];
        if (this.system) {
            this.messages.push({ role: "system", content: system });
        }

        this.client = new OpenAI({
            baseURL: ENDPOINT,
            apiKey: API_KEY
        });
    }

    async call(message: string): Promise<string> {
        this.messages.push({ role: "user", content: message });
        const result = await this.execute();
        this.messages.push({ role: "assistant", content: result });
        return result;
    }

    private async execute(): Promise<string> {
        this.numberOfPromptsSent++;
        console.log(`\n-- Prompt ${this.numberOfPromptsSent} --`);
        console.log("\n=== FULL PROMPT SENT TO MODEL ===");
        
        // Show the complete prompt being sent to the model
        this.messages.forEach((msg, i) => {
            console.log(`  ${msg.content}`);
            console.log();
        });

        const response = await this.client.chat.completions.create({
            model: DEPLOYMENT_NAME!,
            messages: this.messages
        });

        // Uncomment this to print out token usage each time
        // console.log(response.usage);
        
        console.log("=== END OF PROMPT ===\n");
        return response.choices[0].message.content || "";
    }
}

function showExamples(): void {
    console.log("=".repeat(80));
    console.log("CHATBOT EXAMPLES - Simple Q&A");
    console.log("=".repeat(80));
    
    const examples = [
        "What states does Ohio share borders with?",
        "Calculate the square root of 256.",
        "How many r's are in strawbery?",
        "Who was the first president of the United States?"
    ];
    
    examples.forEach((example, i) => {
        console.log(`\n--- Example ${i + 1} ---`);
        console.log(example);
    });
}

async function query(question: string, bot: ChatBot): Promise<string> {
    const result = await bot.call(question);
    return result;
}

async function interactiveLoop(): Promise<void> {
    // Main interactive loop for user queries
    console.log("=".repeat(80));
    console.log("INTERACTIVE CHATBOT - Have a conversation!");
    console.log("Type your questions and press Enter. Use Ctrl+C to exit.");
    console.log("=".repeat(80));
    
    const bot = new ChatBot(); // Create one bot instance to maintain conversation history

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    const askQuestion = (): Promise<string> => {
        return new Promise((resolve) => {
            rl.question("\nYour question: ", (answer) => {
                resolve(answer.trim());
            });
        });
    };

    try {
        while (true) {
            const userQuery = await askQuestion();
            if (userQuery) { // Only process non-empty queries
                const result = await query(userQuery, bot);
                console.log(`Answer: ${result}`);
            } else {
                console.log("Please enter a question.");
            }
        }
    } catch (error) {
        console.log("\n\nGoodbye! Thanks for chatting!");
    } finally {
        rl.close();
    }

    // Handle Ctrl+C gracefully
    process.on('SIGINT', () => {
        console.log("\n\nGoodbye! Thanks for chatting!");
        rl.close();
        process.exit(0);
    });
}

// Main execution
async function main(): Promise<void> {
    // First show examples
    showExamples();
    
    // Then start interactive mode
    await interactiveLoop();
}

if (require.main === module) {
    main().catch(console.error);
}

export { ChatBot, query, showExamples, interactiveLoop };